### ISSUE TITLE
Bug: @typescript-eslint/require-await should not be recommended

### ISSUE THREAD

**Issue by runspired** [typescript-eslint/typescript-eslint#11731]
Recommended rules should represent best practices. At a minimum, using async without await **is not** a bad practice, and in most usages it represents a best practice helping to safeguard the invocation and API surface. The alternative is both more brittle and more verbose, prone to accidentally leaking synchronously thrown errors.

What's more, linting against using recommended language features - especially in a way that changes behavior - is just odd.

**Comment by NullVoxPopuli** [comment]
yeees!

I find the indicator of `async` on a method/function so useful.

Like, especially for boilerplate code, 
```ts
async foo() {
  // TODO: implement this
}
```

this is a signal to consumers to use `await` -- which we lose with `@typescript-eslint/require-await`.

Sure, I "could" add an explicit return type -- but inference is really nice in app dev, ya know?

**Comment by bradzacher** [comment]
> using async without await is not a bad practice, and in most usages it represents a best practice helping to safeguard the invocation and API surface

What does this mean? How does an unnecessary `async` keyword safeguard anything?

> What's more, linting against using recommended language features - especially in a way that changes behavior - is just odd.

But the rule doesn't lint against using it - it lints against using it without the sister feature `await` (or similarly manually returning a promise).

It is linting against what is considered **_by a majority of people_** to be incorrect usage of `async` - as a way to wrap a return value in a promise _implicitly_.

**Comment by bradzacher** [comment]
It looks like you've filed this issue using the bug report template instead of the config change template. I'm not going to close this and force you to re-file it with the correct template but please be mindful of the template you're choosing in future.

Additionally as per our docs 
https://typescript-eslint.io/users/configs#suggesting-configuration-changes

> If you feel strongly that a specific rule should (or should not) be one of these configurations, please file an issue along with a detailed argument explaining your reasoning.


Currently your reasoning boils down to "I don't think using async without await is bad practice" - which is not enough detail, nor is it a compelling e ought argument for us to consider making a change.

**Comment by NullVoxPopuli** [comment]
> But the rule doesn't lint against using it - it lints against using it without the sister feature await (or similarly manually returning a promise).

it's a virus though.


because like how `await` makes you await all your callers, removing `async` will force the `await` to be removed in other locations, which cascades the removal of `async`.

I think the rule only makes sense if combined with another lint rule to require explicit return types.

**Comment by runspired** [comment]
@bradzacher RE what template to use: gate keeping is bad, I appreciate you being flexible: however I did read the docs suggested by the issue template (no relevant things for filing this kind of ticket were in them) and no templates applied. I'm quite dismayed by projects that try to discourage participation by making it extremely hard to open simple issues and/or pipe issues to dev/null as discussions. Expecting me to have found something buried in a massive docs site for a massive project is unreasonable.

> How does an unnecessary async keyword safeguard anything?

There are two massive functional differences between a function marked async and one not. By definition that means it is *not* unnecessary.

First, they always return a promise. It's much easier to convert an async API into a sync one if later it turns out to not have needed to be async, but the inverse is not true. If you design an API as sync, its near impossible to make it async later. For this reason, its a *very* common practice to design APIs that are explicitly async when there's a good chance that later they will need to make use of that.

Second, errors thrown by the function body are always thrown async, In many cases this prevents errors from crashing an application, and even where it doesn't it enables more reliable error management.

> It is linting against what is considered by a majority of people to be incorrect usage of async

I have never seen someone recommend against using `async` in this way, certainly not a majority of people. Moreover I have rarely seen a project not immediately disable this lint rule for its impracticality. I opened this ticket because I got tired of it being one of the first things that needs to be done to get new projects up and running.

> a way to wrap a return value in a promise implicitly.

That seems like an odd definition of implicit, as using a language keyword to do a thing explicitly is in fact extremely explicit.

**Comment by runspired** [comment]
> It is linting against what is considered by a majority of people to be incorrect usage of async

I would even take my feedback on this a step further, if this were true (which it is not), why has the spec committee recently opted to codify this precise usage into parity for promises with `Promise.try` ?

And (barring Promise.try adoption) is it really good to be pushing folks to change

```ts
async function mightThrow() {
  return someWork();
}
```

into (dangerous)

```ts
function mightThrow() {
  return Promise.resolve(someWork())
}
```

or (complicated and different timing)

```ts
function mightThrow() {
  return Promise.resolve().then(someWork)
}
```

or (poorly understood)

```ts
function mightThrow() {
  return new Promise((resolve) => resolve(someWork()));
}
```

**Comment by Josh-Cena** [comment]
I agree there are many good reasons to disable this rule but I disagree it should not be recommended by default. Async is not a zero-cost abstraction. Imagine a library exporting a function `add = async (a, b) => a + b`. Now you force all your consumers to be async—even if they are already async, you are now forcing an `await`, which introduces another reentrancy point and potentially changes the semantics of your program if not just making it slower.

In today's JS, I'm expecting every async operation to be immediately awaited, if not wrapped in `Promise.all` and then awaited, so error handling is exactly the same as sync code. Just lifting your logic into a promise does not make error handling any easier because the caller is going to immediately `await` anyway, and eventually a try...catch is needed. This was a big reason why `Promise.try` was frowned upon when it was proposed, and it was only accepted for feature parity with existing libraries and conceptual consistency with async-await; there's no implicit recommendation from TC39 about how async code should be organized.

This kind of "stub" situation is not unique to this rule: `no-unused-variables`, `no-unnecessary-condition`, etc. all have the same problem. ts-eslint can only assume your code is complete and ready to be used by your consumers as-is and cannot guess what you intend to change in the future.

FYI, for my own ESLint config, I set it to warning, see https://jc-verse.github.io/js-style-guide/eslint-base/async#require-await:

> This is only a warning, because there are many cases where the lack of `await` is intentional:
>
> - The function is a stub and will be implemented later.
> - It's a callback function and the caller will `await` it. You want to remind readers that `await` is allowed in this function.
> - The caller explicitly _requires_ a `Promise` instance to be returned.
> - The function calls another function that is planned to become async in the future.
> - The function is a method that derives from an async base method.
>
> You can use your discretion to disable this rule in these cases.

**Comment by bradzacher** [comment]
> I have never seen someone recommend against using async in this way, certainly not a majority of people.

[The lint rule has been in the recommended set since v2 of our project was released - **_SIX_** years ago.](https://github.com/typescript-eslint/typescript-eslint/pull/729)
This issue is the first one we've ever had asking to remove the rule from the recommended set.

So in six years and with tens of millions of installs each week - you are the first person to actively ask us to remove it.

Now obviously this doesn't directly indicate that there isn't anyone else that aligns with you (as evidenced by the fact that someone else commented on the thread). But it does at the very least indicate that either the vast majority of people agree with its inclusion in the recommended set - or that they at least are apathetic to its inclusion.

**Comment by kristof-mattei** [comment]
I have disabled this rule once of twice at a line based level, both with a comment like 'fire-and-forget'. 

In all the other situations this lint has prevented bugs. 

I vote for keeping it.

**Comment by bradzacher** [comment]
A [quick search using sourcegraph shows that less than 1000 repos explicitly disable the lint rule](https://sourcegraph.com/search?q=context:global+/%40typescript-eslint%5C/require-await%5B%22%27%5D:+%5C%5B%3F%28%5B%22%27%5Doff%7C1%29/&patternType=keyword&sm=0). Which further reinforce the idea that the majority of people are okay with the rule.

Every major release we have also prepared an issue where we explicitly list the changes we are making to the configs. Each major we have listed the rule as remaining and nobody has complained and all maintainers have been in favour of it.

With these three pieces of evidence, I would confidently state that at the very least the majority of people are _at least_ apathetically okay with the rule. Though based on my own anecdotal experience I would still feel confident raising this to "the majority of people agree with the rule".

**Comment by bradzacher** [comment]
So far your arguments against the rule have just been explanations of how async/await work - of which we are very, very aware.

As shown above - the rule is in the recommended set because it is a best-practice for TS code that the majority of people agree with.

That being said - no decision is final and given a compelling argument we're open to changing things.

What we're looking for is clear and evidence of why we should change the status quo.

For example of how it is doing harm as a recommend rule, how the rule's reports do not align with majority case best practices, etc.

Not just some anecdotal explanations - some clear evidence, eg:
- sourcegraph queries showing anti-patterns that are proliferated, large numbers of disable comments, large numbers of turning the rule off in configs)
- a number of well-regarded blogs or tweets against the rule or what it enforces

As you are the one requesting this change, the onus is on you to do the research and convince us to make the change.

**Comment by NullVoxPopuli** [comment]
For me, the most common place it gets in the way is in generated boilerplate -- i maintain a fair number of libraries, and project/file generators ( and many of my consumers use the recommended type-checking lints ).

For example, when generating a boilerplate function:
```js
async request() {
  // TODO: fill this out
  //       and return some object
  return {};
}
```
the linter will complain about this, but the function really is async, and may be called by non-user code and will be awaited anyway. so the async is a good signal to the user that they are allowed to await in here -- but if the lint errors on this right away, it distracts from the purpose of what is trying to be communicated with the sample / boilerplate as folks begin learning new concepts / generating new files.

**Comment by Josh-Cena** [comment]
As I said: starter templates and stubs are often full of lint errors, such as unused variables, unnecessary classes, etc. Many features need accompanying usage to convince the linter that the feature is necessary, which is lacking as expected for starter templates. I don't think "no errors on incomplete code" is a goal.

**Comment by runspired** [comment]
A separate argument then - it's odd that a typescript aware linting project is imposing a stylistic preference completely unrelated to types.

**Comment by runspired** [comment]
@kristof-mattei 

> both with a comment like 'fire-and-forget'.

You are confusing this rule with another which guards against dangling promises. This rule prevents zero bugs, it's stylistic only.

**Comment by bradzacher** [comment]
> is imposing a stylistic preference completely unrelated to types

We do not believe it is a stylistic preference though.

We see it as akin to an unused variable. I.e. Not technically a bug in of itself, but likely indicative of an incomplete change and thus something that should be prevented.

Combined with [`await-thenable`](https://typescript-eslint.io/rules/await-thenable/) (also a recommended rule) it's easy to identify code that can be "de-asyncified" - which is broadly a good thing.

As with all lint rules there are edge cases whereby you may want to intentionally have an async function purely for the promise wrapping. But it had been our belief that that usecase is exactly that - a very rare edge case, rather than it being a common case. Being a rare edge case a disable comment is appropriate.

**Comment by Josh-Cena** [comment]
Folks, I think there are a few presuppositions we need to clean up here.

1. typescript-eslint is not "ESLint but for TypeScript features". It also leverages type information so it's useful even on pure JavaScript, such as this rule.
2. The `recommended` set does not mean "reject bugs". It also encourages tidiness, efficiency, and modernness. Can you really call unused variables a bug? In fact async code without await is a bigger code smell than unused constructs because it actually has a user-facing effect and causes them more troubles.
3. `require-yield` is already recommended by ESLint core, and `require-await` is a core rule, just not recommended because it does have false positives. We fixed the false positives with type information and we believe it is now on par with `require-yield` and deserves to also be recommended. You probably didn't complain about `require-yield` just because you don't use generators that much, but here I adapt Biden: "you can't love a lint rule only when you don't get errors". These two rules have the exact same principle, so if you advocate for the de-recommendation of this rule, then you also have to take the same logic to ESLint core and make them de-recommend `require-yield`.
4. I stand by my original argument: it doesn't make sense to frame coloring as a problem today. Adding and removing async/await is basically zero cost to a developer with basically zero refactoring burden, unlike callbacks or `then`. If the value you get is a promise, then TypeScript forces you to add an await and add async; if the value you get is no longer a promise, then `await-thenable` forces you to remove await, and `require-await` forces you to remove async if possible. These changes don't impact anything else: errors are handled the exact same way (with `try...catch`), and other people calling your code can continue to await (and if they have the same ESLint config they will be prompted to do the same).

Here I address the use cases you brought up.

1. On starter templates. (1) Starter templates already have lint errors. If you control their starter ESLint config, you are free to disable this rule with a comment; if you don't, then be prepared for them to use their company's preset which can give tons more errors. (2) You don't have to start by marking this function as async. You can add a comment saying "can be async", or even not write anything—behaviors have to be learned from documentation anyway, not from the starter code. From a library perspective, the user passing a sync vs. async function makes zero difference because you are going to immediately await the result (perhaps within `Promise.all`) anyway.
2. On preempting an API to become async in the future. (1) Consider not shipping WIP functions if they're not yet usable. (2) If the function is usable today without being async, then it should continue to be, and upgrading it to async should happen in a major version, because users are happier with a sync function—even when their code is already async, adding await introduces a reentrancy opportunity and potentially makes races harder to analyze. (3) If you insist on keeping `async`, you probably have to add a comment anyway informing your colleagues not to remove it. Consider just adding a disable comment alongside. WIP code is intended to be full of lint errors and disable comments.

**Comment by bakkot** [comment]
> why has the spec committee recently opted to codify this precise usage into parity for promises with `Promise.try`?

As a person on the committee, I can maybe speak to that - as I see it the purpose of `Promise.try` is for when you are _already_ working directly with Promises, either because you (for whatever reason) are avoiding the use of `async`/`await` entirely or because you are using combinators like `Promise.all`, _and_ you are calling a callback which you do not know to be either sync or async. This is a valid thing to want, but is pretty unusual outside of certain kinds of library code. Its existence should not be taken as evidence that we expect this to be a common thing in most codebases.

Anyway, speaking only for myself, I think this is a good rule to be on by default. Using `async` without `await` is, in fact, almost always an error. It's true that there are very occasional scenarios where you want to break this rule in non-WIP code, but that's true of all lint rules and is why we have `// eslint-disable-next-line`.

**Comment by runspired** [comment]
> Using async without await is, in fact, almost always an error

Can you show a single case where it is an error? I have yet to see one, this feels like folks belittling developers with a personal objection to async APIs.

This rule is - at best - stylistic. The effect of this rule is that codebases must either litter themselves with lint-ignores, rule disables, and/or fallback to the more verbose promise variants.

This rule is - at worst - setting codebases up to fail by discouraging forward-thinking APIs and encouraging patterns that are easier to mess up.

In my decades of work on large JS codebases, making core APIs defensively async has been the saving grace of most of them. When you don't do this, you inherently lock the app into poorly performing and unscalable synchronous patterns. Adding a single keyword upfront has often saved years of refactoring time later.

**Comment by bakkot** [comment]
> Can you show a single case where it is an error?

Sure. The place it commonly comes up, as discussed above, is during refactoring. Imagine you have something like

```js
async function getAndProcessData(url) {
  let data = await (await fetch(url)).json();
  let filtered = data.filter(whatever);
  // some other stuff, all sync
  return result;
}
```
Later you realize that it makes more sense to pass the data in, and keep the processing logic separate, so you refactor:

```js
async function processData(data) {
  let filtered = data.filter(whatever);
  // some other stuff, all sync
  return result;
}
```
But now you have an `async` when you didn't need one, incurring an extra promise tick and forcing consumers to be async. The `async` keyword here is just a remnant of the old implementation which you forgot to remove when removing the last `await`, not a deliberate choice - in other words, an error.

> The effect of this rule is that codebases must either litter themselves with lint-ignores, rule disables, and/or fallback to the more verbose promise variants.

This has not at all been my experience. I have been using this rule in all of my projects for many years and this has happened to me literally zero times. Every single time this rule has been triggered it's something like the above, where I actually do want to remove the `async`.

> In my decades of work on large JS codebases, making core APIs defensively async has been the saving grace of most of them. When you don't do this, you inherently lock the app into poorly performing and unscalable synchronous patterns. Adding a single keyword upfront has often saved years of refactoring time later.

Well, again, my experience differs.

I believe you about your own experience, but that doesn't mean the rule should be disabled by default. The point of the recommended set is to have a broadly useful set of rules which mostly catch real bugs for most people. This does not mean that they are going to be useful for every single person. Sometimes you're the odd one out where the way you write your code causes a rule to have more false positives than the typical programmer experiences. That's OK. You aren't forced to use the recommended ruleset.

**Comment by runspired** [comment]
> But now you have an async when you didn't need one

I'd argue your example is a very good example of where you should have kept the async keyword. You've already changed your mind on the interface once - what happens when you realize you did in-fact need the async fetch for data?

**Comment by bakkot** [comment]
I put the `async` keyword back? It's not usually hard. In my experience this very rarely happens, though.

I accept that you want to defensively put `async` on functions which are in fact synchronous, but this is very much not standard practice. For almost all JavaScript programmers - not you, but almost anyone else - the fact that this code incurs an unnecessary extra promise tick and forces consumers to be async would be considered an error. It is not something they intended to do and has costs they didn't deliberately decide to pay. It's just an oversight from a refactoring. It is _exactly_ the sort of thing that lint rules are there to catch.

**Comment by runspired** [comment]
> I put the async keyword back?

Moving from async=>sync is easy, moving from sync=>async is incredibly hard. As others have noted - `async` is a forcing function for other program code to either also be async or to handle async. When you unravel that support you box yourself in quickly.

> the fact that this code incurs an unnecessary extra promise tick

ticks are cheap, roughly 2ns. A small cost to pay to save hours, days, months or even years of refactoring later.

**Comment by Josh-Cena** [comment]
That means you have a coding style that's different from our "recommended" practice—that's fine, most people have some kind of quirks in their preferences. It just means our recommended set won't work for you out of the box and you are welcome to publish your own ESLint preset for your own use and for your projects. I, for example, have one because my style is full of quirks.

Some people may want to keep unused parameters because "who knows? I might need it again someday. Better not waste time removing this parameter from all call sites and re-add them later". Some people may want to await a non-thenable because "ticks are cheap and I don't want to remove the await; maybe one day the function will become async again". Many recommended rules are to guard against refactor remnants. It's fine if you want to keep them, but that's not OUR recommended approach.

**Comment by JoshuaKGoldberg** [comment]
The team's position is unanimously not in favor of this proposal. We recognize there are some code styles where it's preferable to add syntactically-unnecessary `async` modifiers to functions to keep them asynchronous. However, in our collective experience, that is a very rare and unusual style counter to how the vast majority of projects work. Adding `async` to a function comes with a cost: of complexity, different function runtime semantics, and a different return type. Doing so unnecessarily is counter to standard community practices as we understand them.
﻿
Just to be clear, we're not against `async`/`await` as a language feature at all. That's not what this issue or the rule are about. Even if we as individuals disliked the language feature (quite the opposite), it's not the typescript-eslint team's place to impart arbitrary personal judgements on code patterns. Our preset configs are our best effort to interpret and help enforce common community best practices, conventions, and patterns that are near-universal positives in common JavaScript/TypeScript code.
﻿
We should also note: it's impossible to make a preset "recommended" config that applies to all valid code styles. If you find yourself disabling/configuring >=1 rule in all your projects, you might want to make a [custom shared ESLint config](https://eslint.org/docs/latest/extend/shareable-configs).
﻿
Because this issue has strong feelings from both sides, adding the `evaluating community engagement` label. tl;dr: we'll leave it open for a bunch of months to give folks a chance to articulate their positions on the matter & share so others can do the same. Docs here: https://typescript-eslint.io/contributing/issues/#evaluating-community-engagement

**Comment by johanrd** [comment]
<img width="600" height="619" alt="Screenshot of React Conf 2025 talk by Ricky Hanlon, React Team @ Meta on 'Async React' where the Slide says 'Async by default' and the caption says 'That's an entirely new way to structure your app'" src="https://github.com/user-attachments/assets/2745f8dd-d415-48f8-b61c-38006c2777ed" />
<br/><br/>

https://youtu.be/B_2E96URooA?t=593

I guess what makes a difference is if your mental model is "this function participates in async coordination infrastructure" or not?

**Comment by JoshuaKGoldberg** [comment]
This issue has been open in `evaluating community engagement` for approximately three months. In that time, there's been one comment. No single comment in favor of the proposal has more than 3 👍s. We'll take that as an indication that, although some folks do prefer what this issue proposes, it is not a community consensus or a hot topic for most. I.e. what https://github.com/typescript-eslint/typescript-eslint/issues/11731#issuecomment-3503154551 suggested.

Closing out as wontfix, but thank you to everyone for the discussion! It was nice to see the different points of view. 💙
